src/state/nfts/createNftsStore.ts

```diff
@@ -0,0 +1,485 @@
+      hasNext: !!data.nftCollections.nextPageKey,
+      total_elements: data.nftCollections.totalCollections,
+    };
+
+    return {
+      collections,
+      nftsByCollection: new Map(),
+      pagination,
+    };
+  } catch (error) {
+    logger.error(new RainbowError('Failed to fetch NFT collections data', error));
+    return EMPTY_RETURN_DATA;
+  }
+};
+
+export type RawCollectionResponse = Awaited<ReturnType<typeof fetchNftData>>;
+
+export const createNftsStore = (address: Address | string) =>
+  createQueryStore<NftsQueryData, NftParams, NftsState>(
+    {
+      fetcher: fetchNftData,
+      onFetched: ({ data, params, set }) => setOrPruneNftsData(data, params, set),
+      cacheTime: time.hours(1),
+      params: {
+        walletAddress: address,
+        limit: PAGE_SIZE,
+        pageKey: null,
+        openCollections: [],
+      },
+      debugMode: IS_DEV,
+      staleTime: STALE_TIME,
+    },
+
+    (set, get) => ({
+      collections: new Map(),
+      openCollections: new Set(),
+      nftsByCollection: new Map(),
+      fetchedPages: {},
+      fetchedCollections: {},
+      pagination: null,
+
+      async fetchNextPage() {
+        if (paginationPromise && paginationPromise.address === address) {
+          return paginationPromise.promise;
+        }
+
+        const { fetchedPages, fetch, getPaginationInfo } = get();
+        const paginationInfo = getPaginationInfo();
+        const nextPageKey = paginationInfo?.pageKey ?? null;
+        const hasNextPage = nextPageKey !== null && paginationInfo?.hasNext;
+
+        if (hasNextPage) {
+          const now = Date.now();
+          const isStale = fetchedPages ? Object.values(fetchedPages).some((fetchedAt: number) => now - fetchedAt > STALE_TIME) : false;
+
+          if (!isStale) {
+            paginationPromise = {
+              address,
+              promise: fetch({ pageKey: nextPageKey, limit: PAGE_SIZE }, { force: true, skipStoreUpdates: true })
+                .then(data => {
+                  if (!data) return;
+                  const { collections, nftsByCollection } = get();
+                  set({
+                    collections: collections ? new Map([...collections, ...data.collections]) : data.collections,
+                    nftsByCollection: nftsByCollection ? new Map([...nftsByCollection, ...data.nftsByCollection]) : data.nftsByCollection,
+                    fetchedPages: { ...fetchedPages, [nextPageKey]: Date.now() },
+                    pagination: data.pagination,
+                  });
+                })
+                .finally(() => (paginationPromise = null)),
+            };
+
+            return paginationPromise.promise;
+          }
+
+          paginationPromise = {
+            address,
+            promise: fetch({ pageKey: null, limit: PAGE_SIZE }, { force: true, skipStoreUpdates: true })
+              .then(data => {
+                if (!data) return;
+                set({
```

Comment:
you may want to have an new AbortController in case this re-runs? could just have a let fetchNextPageController in file scope and then fetchNextPageController?.abort() and before the set here you'd do const currentController = fetchNextPageController and if (!currentController.aborted) set()

Reply:
Would this have adverse effects of creating missing pages?

Reply:
if it always should load the page even if new one comes then no need for this